### ISSUE TITLE
Deferring control of the injectable's lifecycle

### ISSUE THREAD

**Issue by stergiom** [Bonemind/syringe#4]
Is there a way of allowing the wrapped end to manage the lifecycle of the injectable? I'd like to use Syringe with an IoC container layer, allowing for relationships to be defined independently, and in certain cases would like a class (or perhaps a reference to a class) more so than an instance of a class to be injected;

```cr
class Foo
  def initialize
    puts "Init'd."
  end
end

class Bar
  def initialize(@foo : Foo.class)
    2.times { @foo.new }
  end
end

Bar.new(Foo)
```

**Comment by elliotize** [comment]
Are you saying you want `Bar.new` to be that same thing as `Bar.new(Foo)` ?

**Comment by elliotize** [comment]
hmm played around with it. Is this to support injecting a class to replace a defined task?

**Comment by stergiom** [comment]
Yes to `Bar.new` being `Bar.new(Foo)`, or perhaps `Bar.new(->{ Foo.new })` which can then be instantiated as needed via `@foo.call`, something like;

```cr
class Foo
  def initialize
    puts "Init'd."
  end
end

class Bar
  def initialize(@foo : Proc(Foo))
    2.times { @foo.call }
  end
end

Bar.new(->{ Foo.new })
```

..this approach might be better suited for the way Syringe currently resolves dependencies.

**Comment by elliotize** [comment]
I'm not sure of the overall use case but yeah should be doable. I can take a stab at it.

**Comment by elliotize** [comment]
I have made this but it depends on my last PR https://github.com/Bonemind/syringe/commit/7e567a22310edd681316ae9cab44d61d68e6a17a
does that fix your issue?

**Comment by stergiom** [comment]
Just gave this a quick run and it seems to have done it, thanks for this. Wrt the use-case, dealing with throw-away objects mostly, `entities` that would have otherwise been hard coded. I'm making these available via DI to completely separate the association out.
